src/engine/BilinearConstraint.cpp

```diff
@@ -0,0 +1,218 @@
+
+    _b1 = atoi( tokensVec[0].ascii() );
+    _b2 = atoi( tokensVec[1].ascii() );
+    _f = atoi( tokensVec[2].ascii() );
+}
+
+NonlinearFunctionType BilinearConstraint::getType() const
+{
+    return NonlinearFunctionType::BILINEAR;
+}
+
+NonlinearConstraint *BilinearConstraint::duplicateConstraint() const
+{
+    BilinearConstraint *clone = new BilinearConstraint( _b1, _b2, _f );
+    *clone = *this;
+    return clone;
+}
+
+void BilinearConstraint::restoreState( const NonlinearConstraint *state )
+{
+    const BilinearConstraint *bilinear = dynamic_cast<const BilinearConstraint *>( state );
+    *this = *bilinear;
+}
+
+void BilinearConstraint::registerAsWatcher( ITableau *tableau )
+{
+    tableau->registerToWatchVariable( this, _b1 );
+    tableau->registerToWatchVariable( this, _b2 );
+    tableau->registerToWatchVariable( this, _f );
+}
+
+void BilinearConstraint::unregisterAsWatcher( ITableau *tableau )
+{
+    tableau->unregisterToWatchVariable( this, _b1 );
+    tableau->unregisterToWatchVariable( this, _b2 );
+    tableau->unregisterToWatchVariable( this, _f );
+}
+
+void BilinearConstraint::notifyLowerBound( unsigned variable, double bound )
+{
+    ASSERT( variable == _b1 || variable == _b2 || variable == _f );
+
+    tightenLowerBound( variable, bound );
+}
+
+void BilinearConstraint::notifyUpperBound( unsigned variable, double bound )
+{
+    ASSERT( variable == _b1 || variable == _b2 || variable == _f );
+
+    tightenUpperBound( variable, bound );
+}
+
+bool BilinearConstraint::participatingVariable( unsigned variable ) const
+{
+    return variable == _b1 || variable == _b2 || variable == _f;
+}
+
+List<unsigned> BilinearConstraint::getParticipatingVariables() const
+{
+    return List<unsigned>( { _b1, _b2, _f } );
+}
+
+void BilinearConstraint::updateVariableIndex( unsigned oldIndex, unsigned newIndex )
+{
+    ASSERT( oldIndex == _b1 || oldIndex == _b2 || oldIndex == _f );
+
+    ASSERT( !_lowerBounds.exists( newIndex ) && !_upperBounds.exists( newIndex ) &&
+            newIndex != _b1 && newIndex != _b2 && newIndex != _f );
+
+    if ( _lowerBounds.exists( oldIndex ) )
+    {
+        _lowerBounds[newIndex] = _lowerBounds.get( oldIndex );
+        _lowerBounds.erase( oldIndex );
+    }
+
+    if ( _upperBounds.exists( oldIndex ) )
+    {
+        _upperBounds[newIndex] = _upperBounds.get( oldIndex );
+        _upperBounds.erase( oldIndex );
+    }
+
+    if ( oldIndex == _b1 )
+        _b1 = newIndex;
+    if ( oldIndex == _b2 )
+        _b2 = newIndex;
+    else if ( oldIndex == _f )
+        _f = newIndex;
+}
+
+void BilinearConstraint::eliminateVariable( __attribute__( ( unused ) ) unsigned variable,
+                                            __attribute__( ( unused ) ) double fixedValue )
+{
+    throw MarabouError( MarabouError::FEATURE_NOT_YET_SUPPORTED,
+                        "Eliminate variable from a Bilinear constraint" );
+}
+
+bool BilinearConstraint::constraintObsolete() const
+{
+    return false;
+}
+
```

Comment:
Always false is right?

Reply:
For now we do not support turning off the bilinear constraints. Perhaps in the future we could support it.